### ISSUE TITLE
Swarm mode. replicas than wanted

### ISSUE THREAD

**Issue by newnius** [docker/docker#27382]
I started a service (storm-supervisor below) and assigned the replicas to be 3. However it turned out to be 17 running, this list shows the node rejected the task but still started the container.
The docker version is 1.12.2-rc3

```
test@ubuntu-7:~$ docker node ls
ID                           HOSTNAME  STATUS  AVAILABILITY  MANAGER STATUS
1n0vlxe76smi1t4gnpear37gr    ubuntu-8  Ready   Active        
cth31u6fgp4d4juelqgt1794s *  ubuntu-7  Ready   Active        Leader
dv7mgk4fjyhcodvb2i0lnsl01    ubuntu-9  Down    Active        
```

```
test@ubuntu-7:~$ docker service ls
ID            NAME              REPLICAS  IMAGE                                    COMMAND
0gzfp9urn3ke  zookeeper1        1/1       sun/zookeeper                            
24ozqnx7yegv  zookeeper2        1/1       sun/zookeeper                            
27p97mbzzejy  mysql             1/1       mysql:5.6                                
2d5u293rvuoy  storm-supervisor  17/3      sun/storm                                start-service supervisor
```

```
test@ubuntu-7:~$ docker service ps storm-supervisor 
ID                         NAME                    IMAGE      NODE      DESIRED STATE  CURRENT STATE                    ERROR
0ifny4a9m19e3i5gapbcbyi0w  storm-supervisor.1      sun/storm  ubuntu-7  Running        Running 2 days ago               
7lmghokgg3bec3zf6nctv34b8   \_ storm-supervisor.1  sun/storm  ubuntu-9  Shutdown       Shutdown 2 days ago              
dd8t8yugy48n7gg5w2u01j3y2   \_ storm-supervisor.1  sun/storm  ubuntu-8  Shutdown       Rejected 2 days ago              "No such image: sun/storm:late…"
7kbrt7goamiakpheeez8dttcz   \_ storm-supervisor.1  sun/storm  ubuntu-8  Shutdown       Rejected 2 days ago              "No such image: sun/storm:late…"
660hca32708qqhmxjr6w5kikg   \_ storm-supervisor.1  sun/storm  ubuntu-8  Shutdown       Rejected 2 days ago              "No such image: sun/storm:late…"
b3n1c68wiaiwaow0z0emk1v35  storm-supervisor.2      sun/storm            Ready          New 4 seconds ago                
75y6f8snr815mlxftvs7c93xt   \_ storm-supervisor.2  sun/storm            Running        New 8 seconds ago                
33yf727hd1060gqc11f1zlmrd   \_ storm-supervisor.2  sun/storm  ubuntu-8  Running        Rejected less than a second ago  "No such image: sun/storm:late…"
11jjatalgjk9eqnozjdmyoutc   \_ storm-supervisor.2  sun/storm            Running        New 15 seconds ago               
ca0hlq6ghm6qxl52ex4ep3ldj   \_ storm-supervisor.2  sun/storm            Running        New 20 seconds ago               
emezr2phew6hcfw48tqp9punj   \_ storm-supervisor.2  sun/storm            Running        New 22 seconds ago               
0o3pdta4snler710l3laq4wtv   \_ storm-supervisor.2  sun/storm            Running        New 25 seconds ago               
0ybhr8d0tp89t0e8pc78wh2vf   \_ storm-supervisor.2  sun/storm            Running        New 42 seconds ago               
98qf9vfi1xdq74ghiktlpkvva   \_ storm-supervisor.2  sun/storm            Running        New 47 seconds ago               
b7vcmu8h7etgt2aekcn4c7pn4   \_ storm-supervisor.2  sun/storm            Running        New about a minute ago           
bk5hjxqasfitph6ltaqbnpvv8   \_ storm-supervisor.2  sun/storm            Running        New about a minute ago           
3qj5z9s9trs6bjanio84essjr   \_ storm-supervisor.2  sun/storm            Running        New about a minute ago           
17zvbraty2vcs034y92m0ngws   \_ storm-supervisor.2  sun/storm  ubuntu-8  Running        Preparing 3 seconds ago          
3kb5cehztmflm7j0aysk7z5ic   \_ storm-supervisor.2  sun/storm            Running        New about a minute ago           
892kon57zixt15fj8helix8ug   \_ storm-supervisor.2  sun/storm            Running        New 2 minutes ago                
2ml2kzvs4a7er3hozcfjyokvv   \_ storm-supervisor.2  sun/storm            Running        New 2 minutes ago                
esv41pfiu8e6dj0sfn2ox2y27   \_ storm-supervisor.2  sun/storm            Running        New 2 minutes ago                
7rig83vv5qu3o6tkxht0e6ruu   \_ storm-supervisor.2  sun/storm            Running        New 3 minutes ago                
evc9byp39f4cixpdrxusfb24z   \_ storm-supervisor.2  sun/storm            Running        New 3 minutes ago                
7j9kyrhm4mb9w16ala75y39nv   \_ storm-supervisor.2  sun/storm  ubuntu-7  Running        Running 47 hours ago             
4ktwo4yrumeh5miqr0isp7qgz   \_ storm-supervisor.2  sun/storm  ubuntu-7  Running        Running 2 days ago               
2bkpkw9k2c2mtjft4jk00g6bm   \_ storm-supervisor.2  sun/storm  ubuntu-7  Running        Running 2 days ago               
2757j57ehfujdqjbm71f3w1wg   \_ storm-supervisor.2  sun/storm  ubuntu-7  Running        Running 2 days ago               
cyu7cj4owb6vz93kqye1j1ht2   \_ storm-supervisor.2  sun/storm  ubuntu-7  Running        Running 2 days ago               
665f7jy4qm58sgragktki6vus   \_ storm-supervisor.2  sun/storm  ubuntu-7  Running        Running 2 days ago               
7sytyzdtcv60qk3fgl0962k7q   \_ storm-supervisor.2  sun/storm  ubuntu-7  Running        Running 2 days ago               
71lud0ryauickt4d2ygci0lkw   \_ storm-supervisor.2  sun/storm  ubuntu-7  Running        Running 2 days ago               
2gnjxtlp46w4izd06wcd06lk3   \_ storm-supervisor.2  sun/storm  ubuntu-7  Running        Running 2 days ago               
dyw0rilas0cw1siv69tyc13dm   \_ storm-supervisor.2  sun/storm  ubuntu-7  Running        Running 2 days ago               
cg7xt5725zxvanarb36v97jxx   \_ storm-supervisor.2  sun/storm  ubuntu-7  Running        Running 2 days ago               
2hy2kivzdvbqweabt0hx6isvo   \_ storm-supervisor.2  sun/storm  ubuntu-7  Running        Running 2 days ago               
erlo24b80scerc16owtty292m   \_ storm-supervisor.2  sun/storm  ubuntu-7  Running        Running 2 days ago               
95ppduiq0429a5pihpp9jw3fw   \_ storm-supervisor.2  sun/storm  ubuntu-7  Running        Running 2 days ago               
dzliaka2pmdk53tdoyl1nthk1   \_ storm-supervisor.2  sun/storm  ubuntu-7  Running        Running 2 days ago               
b45erxbuzkc8a6pjwd6kb6b78  storm-supervisor.3      sun/storm  ubuntu-7  Running        Running 2 days ago   
```            

```
docker service create \
--replicas ${SUPERVISOR_NUM} \
--name supervisor \
--network my-network \
--mount type=bind,src=/mnt/Package,dst=/package,readonly \
--mount type=bind,src=/mnt/Data/storm-logs,dst=/logs \
31z4/storm:1.0.2 \
storm supervisor -c storm.zookeeper.servers="${storm_zookeeper_str}" -c storm.log.dir="/logs" -c nimbus.seeds="${nimbus_seeds_str}"
```

**Comment by aluzzardi** [comment]
@newnius Could you please provide exact repro steps, including **all** services (mysql, zookeeper1, zookeeper2, mysql and storm-supervisor)?

/cc @aaronlehmann

**Comment by newnius** [comment]
It's not easy to reproduce this - it depends on time.
possible steps to reproduce this: 
1. start 3 zookeepers
```
ZOOKEEPER_NUM=3
zookeeper_str=""
if [ $ZOOKEEPER_NUM -gt 0 ]; then
  zookeeper_str+="zookeeper_1"
fi

for i in $( seq 2 $ZOOKEEPER_NUM )
do
  zookeeper_str+=",zookeeper_"${i}
done
for i in $( seq 1 $ZOOKEEPER_NUM )
do
docker service create \
-e MYID=${i} \
-e SERVERS=${zookeeper_str} \
--replicas 1 \
--network my-network \
--endpoint-mode dnsrr \
--name=zookeeper_${i} mesoscloud/zookeeper:3.4.8-ubuntu-14.04
done
```


2. start 3 storm nimbus
```
ZOOKEEPER_NUM=3
NIMBUS_NUM=3

zookeeper_str=""
storm_zookeeper_str=""
nimbus_seeds_str=""
if [ $ZOOKEEPER_NUM -gt 0 ]; then
  zookeeper_str+="zookeeper_1"
  storm_zookeeper_str+="[\"zookeeper_1\""
fi

for i in $( seq 2 $ZOOKEEPER_NUM )
do
  zookeeper_str+=",zookeeper_"${i}
  storm_zookeeper_str+=",\"zookeeper_${i}\""
done

storm_zookeeper_str+="]"                                                                                                                                                    
                                                                                                                                                                      
if [ $NIMBUS_NUM -gt 0 ]; then                                                                                                                                        
  nimbus_seeds_str+="[\"storm_nimbus_1\""                                                                                                                                                  
fi                                                                                                                                                                    
                                                                                                                                                                      
for i in $( seq 2 $NIMBUS_NUM )                                                                                                                                       
do
  nimbus_seeds_str+=",\"storm_nimbus_${i}\""
done
nimbus_seeds_str+="]"
for i in $( seq 1 $NIMBUS_NUM )
do
docker service rm storm_nimbus_${i}
docker service create \
--replicas 1 \
--name storm_nimbus_${i} \
--network my-network \
--endpoint-mode dnsrr \
--mount type=bind,src=/mnt/Package,dst=/package,readonly \
31z4/storm:1.0.2 \
storm nimbus -c storm.zookeeper.servers="${storm_zookeeper_str}" -c nimbus.seeds="${nimbus_seeds_str}" 
done
```

3. start supervisor
```
ZOOKEEPER_NUM=3
NIMBUS_NUM=3
SUPERVISOR_NUM=5

zookeeper_str=""
storm_zookeeper_str=""
nimbus_seeds_str=""
zookeeper_str+="zookeeper_1"
storm_zookeeper_str+="[\"zookeeper_1\""

for i in $( seq 2 $ZOOKEEPER_NUM )
do
  zookeeper_str+=",zookeeper_"${i}
  storm_zookeeper_str+=",\"zookeeper_${i}\""
done

storm_zookeeper_str+="]"
nimbus_seeds_str+="[\"storm_nimbus_1\""

for i in $( seq 2 $NIMBUS_NUM )
do
  nimbus_seeds_str+=",\"storm_nimbus_${i}\""
done
nimbus_seeds_str+="]"

docker service create \
--replicas ${SUPERVISOR_NUM} \
--name storm_supervisor \
--network my-network \
--mount type=bind,src=/mnt/Package,dst=/package,readonly \
--mount type=bind,src=/mnt/Data/storm-logs,dst=/logs \
31z4/storm:1.0.2 \
storm supervisor -c storm.zookeeper.servers="${storm_zookeeper_str}" -c storm.log.dir="/logs" -c nimbus.seeds="${nimbus_seeds_str}"
```

And
4. wait......

The dns record (or route rules?) seems being updated to other nodes after a delay which is longer than supervisor could find zookeeper or nimbus, so it keeps failing, again and again. Then there may be more replicas than wanted.
Now I use another image which keeps retrying to avoid this situation.
I guess this is caused by worker returns earlier than it really shuts down the failed container.

**Comment by aaronlehmann** [comment]
I believe this is fixed by https://github.com/docker/swarmkit/pull/1305

I've flagged that PR for backport to 1.12.3.

**Comment by aluzzardi** [comment]
1.12.3?

**Comment by aaronlehmann** [comment]
This is fixed in 1.12.3.
